server/app/services/applications/ProgramAdminApplicationService.java

```diff
@@ -239,4 +217,24 @@ public Optional<String> getNote(Application application) {
         .findFirst()
         .map(app -> app.getDetails().noteEvent().get().note());
   }
+
+  /**
+   * Retrieves the application with the given ID and validates that it is associated with the given
+   * program.
+   */
+  public Optional<Application> getApplication(long applicationId, ProgramDefinition program) {
+    return validateProgram(
+        applicationRepository.getApplication(applicationId).toCompletableFuture().join(), program);
+  }
+
+  /** Validates that the given application is part of the given program. */
+  private Optional<Application> validateProgram(
+      Optional<Application> application, ProgramDefinition program) {
+    if (application.isEmpty()
+        || application.get().getProgramName().isEmpty()
+        || !application.get().getProgramName().equals(program.adminName())) {
+      return Optional.empty();
+    }
+    return application;
+  }
```

Comment:
**issue (blocking):** Returning an empty optional is not an obvious way of indicating that a given piece of data or a given object is invalid. Better alternatives would be to either throw an exception or return an error object such as https://github.com/civiform/civiform/blob/main/server/app/services/ErrorAnd.java